### PR TITLE
Add Teams custom cmdlets for Pre-Approval

### PR DESCRIPTION
**Pull request by subray2014**
﻿

Fixes #

### Changes proposed in this pull request

Identity has recently exposed APIs to create pre-approval permission grant policies which allow admins to configure whether specific apps can be granted a set of permissions based on sensitivity of the involved data. Since these policies operate on an AAD app level, the Teams workload wants to build a TeamsApp-centric abstraction on top of these policies in collaboration with Identity. Teams is planning to launch a set of cmdlets which provide a simple experience to admins to pre-approve RSC permissions specific to Teams.  
The current set of APIs developed by Identity operate on a set of identifiers different from the Teams ecosystem. The preapproval policy APIs work on AAD apps. In the Teams Admin Portal, admins operate on Teams Apps. They approve/block or setup policies on Teams Apps. It is expected that the RSC preapproval operation will be applied in a recursive manner on all components associated with a Teams App such as the AAD app backing it. We need a way to bridge this gap and provide admins with a way of approving RSC for Teams Apps.  

This change adds a set of custom cmdlets to the Teams workload:
1. New-MgBetaTeamsAppPreapproval - Perform the set of steps needed to create a preapproval policy and associate the AAD app backing a Teams App with it.
Params:
a. TeamsAppId - Id of the app being approved.
b. ResourceSpecificApplicationPermissionsAllowedForTeams - Permissions allowed to the Teams App when installed in teams once this instance has been created.
c. ResourceSpecificApplicationPermissionsAllowedForChats - Permissions allowed to the Teams App when installed in chats once this instance has been created.
d. TeamLevelSensitivityLabelCondition - Condition for whether sensitivity labels must be applied on teams for this approval to take effect.
e. SpecificSensitivityLabelIdsApplicableToTeams - If TeamLevelSensitivityLabelCondition required specific labels, then this param identifies those label Ids. 

3. Get-MgBetaTeamsAppPreapproval - Get preapproval policy associated with the AAD app backing a Teams App. This cmdlet will output a custom object which flattens the preapproval policy and provides a user friendly view to the admin.  
Params:
a. TeamsAppId - Id of the Teams app

4. Update-MgBetaTeamsAppPreapproval - Perform the set of steps needed to update preapproval policy associated with the AAD app backing a Teams App. 
Params:
a. TeamsAppId - Id of the Teams app.
b. ResourceSpecificApplicationPermissionsAllowedForTeams - Permissions allowed to the Teams App when installed in teams once this instance has been created.
c. ResourceSpecificApplicationPermissionsAllowedForChats - Permissions allowed to the Teams App when installed in chats once this instance has been created.
d. TeamLevelSensitivityLabelCondition - Condition for whether sensitivity labels must be applied on teams for this approval to take effect.
e. SpecificSensitivityLabelIdsApplicableToTeams - If TeamLevelSensitivityLabelCondition required specific labels, then this param identifies those label Ids. 

5. Remove-MgBetaTeamsAppPreapproval - Remove the association between a Teams App and a pre approval policy and delete the policy.
Params:
a. TeamsAppId - Id of the Teams app

-
-
-

### Other links

-
-
-

### CODE REVIEW

File: src/Teams/beta/custom/GetMgTeamsAppPreApproval_Get.cs [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+    using static Microsoft.Graph.Beta.PowerShell.Runtime.Extensions;
+
+    /// <summary>
+    /// Get Preapproval policy associated with a Teams App.
+    /// </summary>
+    [global::System.Management.Automation.Cmdlet(global::System.Management.Automation.VerbsCommon.Get, @"MgBetaTeamsAppPreApproval_Get")]
+    [global::System.Management.Automation.OutputType(typeof(Microsoft.Graph.Beta.PowerShell.Models.IMicrosoftGraphTeamsAppPreApproval))]
+    [global::Microsoft.Graph.Beta.PowerShell.Description(@"Get Preapproval policy associated with a Teams App.")]
+    [global::Microsoft.Graph.Beta.PowerShell.Generated]
+    public partial class GetMgTeamsAppPreApproval_Get : global::System.Management.Automation.PSCmdlet,
```
Comment (peombwa) [inline src/Teams/beta/custom/GetMgTeamsAppPreApproval_Get.cs:15]: Let's change the file and class name to `GetMgBetaTeamsAppPreApproval` to match the name convention of beta commands.

Comment (subray2014) [reply src/Teams/beta/custom/GetMgTeamsAppPreApproval_Get.cs]: Renamed

File: src/Teams/beta/custom/UpdateMgTeamsAppPreApproval_Update.cs [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+    using static Microsoft.Graph.Beta.PowerShell.Runtime.Extensions;
+
+    /// <summary>
+    /// Update Preapproval policy associated with a Teams App.
+    /// </summary>
+    [global::System.Management.Automation.Cmdlet(global::System.Management.Automation.VerbsData.Update, @"MgBetaTeamsAppPreApproval_Update", SupportsShouldProcess = true)]
+    [global::System.Management.Automation.OutputType(typeof(bool))]
+    [global::Microsoft.Graph.Beta.PowerShell.Description(@"Update Preapproval policy associated with a Teams App.")]
+    [global::Microsoft.Graph.Beta.PowerShell.Generated]
+    public partial class UpdateMgTeamsAppPreapproval_Update : global::System.Management.Automation.PSCmdlet,
```
[the remark this answers is not in the document]
Comment (subray2014) [reply src/Teams/beta/custom/UpdateMgTeamsAppPreApproval_Update.cs]: Renamed

File: src/Teams/beta/custom/RemoveMgTeamsAppPreApproval_Delete.cs [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+    using static Microsoft.Graph.Beta.PowerShell.Runtime.Extensions;
+
+    /// <summary>
+    /// Delete PreApproval policy associated with a Teams App.
+    /// </summary>
+    [global::System.Management.Automation.Cmdlet(global::System.Management.Automation.VerbsCommon.Remove, @"MgBetaTeamsAppPreApproval_Delete", SupportsShouldProcess = true)]
+    [global::System.Management.Automation.OutputType(typeof(bool))]
+    [global::Microsoft.Graph.Beta.PowerShell.Description(@"Delete Preapproval policy associated with a Teams App.")]
+    [global::Microsoft.Graph.Beta.PowerShell.Generated]
+    public partial class RemoveMgTeamsAppPreApproval_Delete : global::System.Management.Automation.PSCmdlet,
```
[the remark this answers is not in the document]
Comment (subray2014) [reply src/Teams/beta/custom/RemoveMgTeamsAppPreApproval_Delete.cs]: Did you mean RemoveMgBetaTeamsAppPreApproval? Remove is the verb used in all the other "delete" cmdlets.

[the remark this answers is not in the document]
Comment (peombwa) [reply src/Teams/beta/custom/RemoveMgTeamsAppPreApproval_Delete.cs]: @subray2014, yes, I meant `RemoveMgBetaTeamsAppPreApproval`. I can see you already performed the rename. I'll review the PR one last time.

**Review by peombwa — COMMENTED** [review]
Please singularize `Teams` in the command names to be consistent with existing commands:
image

`*-MgBetaTeamsAppPreApproval` -> `*-MgBetaTeamAppPreApproval`.

**Review by peombwa — APPROVED** [review]

**Pull request merged**